ddprof-ffi/src/exporter.rs

```diff
@@ -277,6 +283,51 @@ pub unsafe extern "C" fn profile_exporter_send(
     }
 }
 
+fn unwrap_cancellation_token<'a>(
+    cancel: Option<NonNull<CancellationToken>>,
+) -> Option<&'a tokio_util::sync::CancellationToken> {
+    cancel.map(|c| {
+        let wrapped_reference: &CancellationToken = unsafe { c.as_ref() };
+        let unwrapped_reference: &tokio_util::sync::CancellationToken = &(*wrapped_reference).0;
+
+        unwrapped_reference
+    })
+}
+
+/// Can be passed as an argument to send and then be used to asynchronously cancel it from a different thread.
+#[no_mangle]
+#[must_use]
+pub extern "C" fn ddprof_ffi_CancellationToken_new() -> *mut CancellationToken {
+    Box::into_raw(Box::new(CancellationToken(
+        tokio_util::sync::CancellationToken::new(),
+    )))
+}
+
+/// Cancel send that is being called in another thread with the given token.
+/// Note that cancellation is a terminal state; cancelling a token more than once does nothing.
+/// Returns `true` if token was successfully cancelled.
+#[no_mangle]
+pub extern "C" fn ddprof_ffi_CancellationToken_cancel(
+    cancel: Option<NonNull<CancellationToken>>,
+) -> bool {
+    let cancel_reference = match unwrap_cancellation_token(cancel) {
+        Some(reference) => reference,
+        None => return false,
+    };
+
+    if cancel_reference.is_cancelled() {
+        return false;
+    }
+    cancel_reference.cancel();
+
+    true
+}
```

Comment:
This is a race condition, yeah? Some other thread could have cancelled between the check and the cancel. Can we call cancel even if it's been cancelled? If hope so, and if so I'd prefer that and return nothing instead of `bool`.

Reply:
Yeah, looking at the code, cancelling twice doesn't do anything. It's just an atomically updated flag.

Reply:
Yup, this is inherently a race -- the early return is not an attempt at cancelling exactly once (the `CancellationToken` already does that as Paul pointed out), I'm just trying to add some visibility to the caller.

Happy to change this to not return anything, if we feel like this return is not useful.

Reply:
I don't think I would use it, but if you think you would in Ruby for some diagnostic or something go ahead.